### PR TITLE
Format tuple like structs

### PR DESCRIPTION
**Pull request by marcusklaas**
This adds formatting for tuple like structs, and should address https://github.com/nrc/rustfmt/issues/86.

### CODE REVIEW

File: tests/idem/structs.rs [hunk 1]
```diff
@@ -14,6 +14,18 @@ pub struct Foo {
...
+
+struct Qux<'a,
+           N: Clone + 'a,
+           E: Clone + 'a,
+           G: Labeller<'a, N, E> + GraphWalk<'a, N, E>,
+           W: Write + Copy>
+    (AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA, // Comment
+     BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB,
+     G,
+     W);
```
Comment (nrc) [inline tests/idem/structs.rs]: I think I would format like this:

```
struct Qux<'a,
           N: Clone + 'a,
           E: Clone + 'a,
           G: Labeller<'a, N, E> + GraphWalk<'a, N, E>,
           W: Write + Copy>(
    AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA, // Comment
    BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB,
    G,
    W,
);
```

Note the trailing comma - we probably want three options for trailing commas in tuples - always, never, and vertical only (maybe we have these options already, actually).

Comment (marcusklaas) [inline tests/idem/structs.rs]: That looks good, but isn't it inconsistent with the rest of the formatting rules? We usually don't have a 'dangling' closing paren for tuple-like structures.

Comment (nrc) [inline tests/idem/structs.rs]: I don't think we can be consistent with both the expression form and regular structs. I think being closer to other items is more important here (and gives less weird indenting). Struct literals also vary a bit from struct definitions, so I think its OK.

Comment (marcusklaas) [inline tests/idem/structs.rs]: Alright, will implement.

Comment (nrc) [inline tests/idem/structs.rs]: Thanks!

**Comment by nrc** [comment]
Also, if you take this approach I wonder if you can format each field independently so we can cope with a mix of struct-like and tuple-like fields. Obvs whether you use `{}` or `()` will have to be determined by the first field, but the rest could be a mix. That should give slightly nicer code and should be more future proof too.

**Comment by marcusklaas** [comment]
This was my initial approach, but it didn't work out for me. Especially switching between horizontal mode and vertical mode is unique to the tuple variant. It also requires distinct formatting code, and then we'd probably just reimplement `write_list`.. I am of course open to suggestions on how to do this nicely!

**Comment by nrc** [comment]
I might be missing the subtleties having not written the code. But it seems like the vertical vs horizontal is just a parameter to write_list and can be deduced at the same time as whether to use {} or (). Formatting each field into a string should be independent of any formatting of the whole struct.

**Comment by marcusklaas** [comment]
Right, I think I understand now. So we'd make `visit_field` return a `String` instead of pushing changes directly and then use `write_list` to format the struct in both cases?

**Comment by nrc** [comment]
Yeah, I think that should work. I might be wrong because of attributes on fields, but hopefully not.

**Comment by marcusklaas** [comment]
I'll look into it!

**Comment by marcusklaas** [comment]
Should be functionally correct now, but still needs a refactor. WIP.

**Comment by marcusklaas** [comment]
I am a bit stuck here. What gives me the most troubles is that in a struct like
```rust struct Foo(SomeType, OtherType); ```.
the span for the last `StructField` contains the closing parent.. But not when it is followed by a comment:
```rust struct Foo(SomeType, OtherType /* this is a type */); ```.
Is this something that we should consider a bad span and fix in libsyntax, or is this something we should hack around?

**Comment by nrc** [comment]
@marcusklaas definitely a libsyntax bug, lets fix it there.

**Comment by marcusklaas** [comment]
I don't know what changed, but that bug that I mentioned earlier no longer seems to be present! Still, I have troubles finishing this PR.

Quick recap: I'm trying to format all structs using the `write_list` function. This works fairly well, but there's some pain points. For example, it's not clear how to deal with comments preceding fields:
```rust
struct Problematic {
    // Our first and
    pub x: String, // only field
}
```
This is very difficult, I think. I do have workarounds in mind, but I thought I'd run them by you first. We could pass two associated comments for every item we give to `write_list`: a preceding and a trailing one. We could also "merge" the preceding with the comment with the actual item before we pass them to `write_list`. I like the latter better, but it gives `write_list` fewer options to refactor things. Consider for example this struct:
```rust
struct Tuple(
    // Some int
    i32
);
```
That would ideally be formatted as 
```rust
struct Tuple(i32 /* Some int */);
```
On a related note, I believe the comments we pass to `write_list` should be stripped of "comment tokens" (`//`, `/*` and `*/`) so that'd do such transformations more easily.

This may all be going a bit far, so maybe it's best to just do something that works now (merge preceding comment and item texts) and leave improvements for later. What do you think, @nrc?

**Comment by nrc** [comment]
Hmm, it is a tricky issue. I think there are a few questions here. The first one is, do we want to move comments around? I had been ducking this issue previously. We could imagine having a style guideline that says "all comments on fields should happen on the previous line" or something. My intuition is that, in general, the programmer probably knows best with comments and we should do the minimum of reorganisation. We can certainly re-visit this later though.

I do think we can and should change the kind of comment - i.e., `/* ... */` to `//...` where possible. I don't feel this is high priority, but if it works, then great!

The next question is how much we want to get right. There are already some cases we don't even try to handle, e.g., `pub x /* a comment */: String,` will just drop the comment on the floor. I think this is OK for now, though not ideal in the long run. I think we probably need to handle comments in both the positions you mention in your comment (i.e., pre and post).

Given that there is so much variety in how we could place comments, I believe we have to let write_list do it, rather than trying to do some merging before passing to write_list.

So, all of this suggests that having pre- and post-comments passed to write_list is a good idea. Probably we should pass one list of structs, rather than three lists. Probably need to start using Options rather than empty strings too. I think passing the comments without the // or /* */ is a good idea. (The complicated aspect here is that where to put the comments and what tokens to use depends on how write_list lays them out).

I think this shouldn't be too hard. I think the quick and dirty approach will actually cause problems in the short run too, so isn't worth it. I'd be fine with not covering all the possible situations in write_list for now. I.e., panicking in some situations and fixing them later. Only implement what you need here and to not cause regressions.

The other option if you think that is too slow would be to not handle side comments for now (i.e., only handle comments on the preceding line).

**Comment by marcusklaas** [comment]
Thank you for your wisdom Nicholas. I fully agree and will try to implement what was suggested.

**Comment by marcusklaas** [comment]
I've given this an update. It should close https://github.com/nrc/rustfmt/issues/86 and https://github.com/nrc/rustfmt/issues/76, since struct-like enum variants are formatted now as well. Some progress on https://github.com/nrc/rustfmt/issues/27 has been made as well. Formatting the comments correctly was the biggest challenge, after all.

A brief description of the approach: let `write_list` do most of the heavy lifting. It takes a lift of `ListItem`s now, which is basically a `String` with optional pre- and post-comments, as you suggested. I've defined a pre-comment to be the first comment after an item and additionally that *starts on the same line*. I think that makes sense, but there is room for ambiguity. For example:

```rust
struct AmbiguousIsMe {
    pub x: 32, // this is clearly belongs to x
          // but what about this?
    // this clearly belongs to y
    y: PrivateType
}
```

Most expressions and items that use `write_list` now support pre-comments too! See the `fn-simple.rs` tests. 

There's a million edge cases, which tend to make things messy. For example, a long post-comment will be converted into a block-style comment. The last post-comment in expression-like lists will also be in block-style, to prevent the comments from commenting out actual code:
```rust
fn test(
    arg_one: i32, // looooooooooong comment
    arg_two: &str // comment
) {
```
may become
```rust
fn test(arg_one: i32, // looooooooooong comment
        arg_two: &str /* comment */) {
```
There's many changes here and I am confident about only few of those, so I would be very grateful for any feedback!

File: src/string.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                      // Remove first whitespace, because we'll add that extra later
+                      line.trim()
+                          .trim_left_matches("// ")
+                          .trim_left_matches("//")
+                          .trim_left_matches("/* ")
+                          .trim_left_matches("/*")
+                          .trim_left_matches("* ")
+                          .trim_left_matches('*')
+                          .trim_right_matches("*/")
+                          .trim_right_matches(' ')
```
Comment (cassiersg) [inline src/string.rs]: I think there is a problem with comments like
```rust
// // /* <- yes, I want to talk about comments in my comment
```

File: src/string.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                  });
+
+    s.push_str(closer);
+
+    // Trim inner comment when it fits on one line
+    if s.len() < width && !s.contains('\n') {
+        s.trim().to_owned()
+    } else {
+        s
+    }
```
Comment (cassiersg) [inline src/string.rs]: In which case is this needed ? It seems to me that `s` is already trimmed since all lines are individually trimmed.

File: src/items.rs [hunk 4]
```diff
@@ -264,85 +270,125 @@ impl<'a> FmtVisitor<'a> {
...
+                None => next_span_start
+            };
+            let post_snippet = self.snippet(codemap::mk_sp(get_hi(&item),
+                                                           next_start));
+
+            let comment_end = match new_it.peek() {
+                Some(..) => {
+                    // Block-style post-comment
+                    if before(&post_snippet, "/*", "\n") {
+                        cmp::max(post_snippet.find("*/").unwrap() + 2,
```
Comment (cassiersg) [inline src/items.rs]: Isn't here a problem when `post_snippet` is like `, /* /* */ */` ?

File: src/items.rs [hunk 5]
```diff
@@ -447,110 +497,207 @@ impl<'a> FmtVisitor<'a> {
...
-    {
-        let header_str = self.format_header("struct", ident, vis);
-        self.changes.push_str_span(span, &header_str);
+        let generics_str = match generics {
+            Some(g) => self.format_generics(g,
+                                            opener,
+                                            offset + header_str.len(),
+                                            codemap::mk_sp(span.lo,
+                                                           struct_def.fields[0].span.lo)),
+            None => " {".to_owned()
```
Comment (cassiersg) [inline src/items.rs]: Shouldn't be `opener` instead of `"{"` ?

File: src/lists.rs [hunk 6]
```diff
@@ -130,24 +159,50 @@ pub fn write_list<'b>(items: &[(String, String)], formatting: &ListFormatting<'b
...
-        if tactic == ListTactic::Vertical && comment.len() > 0 {
-            if !comment.starts_with('\n') {
-                result.push(' ');
-            }
-            result.push_str(comment);
+        if tactic == ListTactic::Vertical && item.post_comment.is_some() {
+            let width = formatting.v_width - item_width - formatting.separator.len();
+            let offset = formatting.indent + item_width + formatting.separator.len();
+            let comment = item.post_comment.as_ref().unwrap();
+            // Use block-style only for the last item or multiline comments
```
Comment (cassiersg) [inline src/lists.rs]: According to the [guidelines](http://doc.rust-lang.org/nightly/style/style/comments.html), multiline comments should use line comments instead of block comments, but we're indeed not in a good situation if we have multiline post-comments.

**Comment by cassiersg** [comment]
Except theses details, seems very nice.
I just needed something like this to work on comment handling into expressions.

File: src/string.rs (hunk 2 above)
Comment (marcusklaas) [inline src/string.rs]: Good catch!

File: src/string.rs (hunk 3 above)
Comment (marcusklaas) [inline src/string.rs]: The individual lines are only trimmed right (to preserve indentation within comments). You are right that a `.trim_left()` would suffice here.

File: src/items.rs (hunk 4 above)
Comment (marcusklaas) [inline src/items.rs]: I didn't know that was a valid comment! That is a problem, then.

File: src/lists.rs (hunk 6 above)
Comment (marcusklaas) [inline src/lists.rs]: Yup.. it's not ideal. I welcome suggestions ;-)

File: src/items.rs (hunk 5 above)
Comment (marcusklaas) [inline src/items.rs]: Most definitely!

**Comment by marcusklaas** [comment]
Thank you very much for your excellent commentary, @cassiersg. I'll get to work!

File: src/string.rs (hunk 3 above)
Comment (cassiersg) [inline src/string.rs]: Doesn't `s` start with `opener` ?

Comment (marcusklaas) [inline src/string.rs]: You are absolutely right!

File: src/lists.rs (hunk 6 above)
Comment (cassiersg) [inline src/lists.rs]: When it's not the last item or not in expression context, you could just use inline style (`//`), even if the comment is multiline or exceeds the with. From what I've seen, `rewrite_comment` handle it right.

Comment (marcusklaas) [inline src/lists.rs]: That's true! The problem is that the next time you run `rustfmt` on it, the result won't be the same. Only the first line will remain, and the rest will be viewed as pre-comment for the next item ;-)

Comment (cassiersg) [inline src/lists.rs]: Yes, indeed. We have thus at least three options :
1. use bloc comment
2. detect logical blocs of comments based on indentation
3. move such comments to be pre-comments

In my opinion, the best visually is to move to pre-comments, but it would merge this comment with all pre-comments, but it's not *reversible*, ie if the line before the comment gets shorter, the next run of rustfmt will not place back the comment as post-comment. Probably not a big issue.

Option 2 is a bit (or lot ?) more work, for something that will anyway looks bad. I think it's fair to leave things in a non guidelines-conforming state when the places of the comments are weired, if we don't make things worse.
Option 1 being more conservative than option 3. I think you did it right. The decision to move comments around is probably better to discuss in another PR.

Comment (nrc) [inline src/lists.rs]: I think that in the case of a post-comment, where the user has used /* */ and it is a multiline comment, then it is OK to leave it as a /* */ comment and don't convert to //.

I think we should strongly avoid turning post-comments into pre-comments, I'd rather err on the side of sub-optimal formatting than do that.

I'm also a bit wary about moving post-comments to pre-comments. We should perhaps do this at some point, but I'm wary about diving right in - when we do formatting which is more than just changing whitespace, I think there is a higher bar for 'correctness'.

I also like the invariant (which had not previously occurred to me) that the output of rustfmt should be idempotent under further reformatting.

File: src/string.rs [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Format comments and string literals
```
Comment (nrc) [inline src/string.rs]: I would break this into separate modules for stings and comments, unless there is a lot of overlap I'm not noticing

File: src/lists.rs [hunk 8]
```diff
@@ -130,24 +159,50 @@ pub fn write_list<'b>(items: &[(String, String)], formatting: &ListFormatting<'b
...
+                              comment.trim().len() > width;
+
+            let formatted_comment = rewrite_comment(comment,
+                                                    block_style,
+                                                    width,
+                                                    offset);
+
+            result.push(' ');
+            result.push_str(&formatted_comment);
         }
```
Comment (cassiersg) [inline src/lists.rs]: The handling of comments in `Mixed` mode seems to be incorrect : `line_len` should be updated with the comments len.

Comment (marcusklaas) [inline src/lists.rs]: Excellent observation, thanks!

File: src/comment.rs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+use string::{StringFormat, rewrite_string};
+use utils::make_indent;
+
+pub fn rewrite_comment(orig: &str, block_style: bool, width: usize, offset: usize) -> String {
+    let s = orig.trim();
+
+    // Edge case: block comments. Let's not trim their lines (for now).
+    let opener = if block_style { "/* " } else { "// " };
+    let closer = if block_style { " */" } else { "" };
+    let line_start = if block_style { "* " } else { "// " };
```
Comment (nrc) [inline src/comment.rs]: should this be " * " so the *s align?

Comment (marcusklaas) [inline src/comment.rs]: It's nicer that way, yes. I previously aligned them by increasing the offset by one.

**Comment by marcusklaas** [comment]
Should also fix https://github.com/nrc/rustfmt/issues/87 now.

File: src/lists.rs [hunk 10]
```diff
@@ -41,11 +47,38 @@ pub struct ListFormatting<'a> {
     pub h_width: usize,
     // Available width if we layout vertically
     pub v_width: usize,
+    // Non-expressions, e.g. items, will have a new line at the end of the list.
+    // Important for comment styles.
+    pub is_expression: bool
```
Comment (nrc) [inline src/lists.rs]: Could this be named to be more about what you're doing rather than why? ends_with_newline or ends_block or something? (also needs a trailing comma)

File: src/lists.rs [hunk 11]
```diff
@@ -125,30 +163,156 @@ pub fn write_list<'b>(items: &[(String, String)], formatting: &ListFormatting<'b
...
             _ => {}
         }
 
-        result.push_str(item);
+        // Pre-comments
+        if let Some(ref comment) = item.pre_comment {
+            result.push_str(&rewrite_comment(comment,
+                                             // Block style in non-vertical mode
+                                             tactic != ListTactic::Vertical,
+                                             1000,
```
Comment (nrc) [inline src/lists.rs]: I'm not a fan of these magic 1000s - could you use an Option<usize> rather than a usize?

File: src/lists.rs (hunk 10 above)
Comment (nrc) [inline src/lists.rs]: The more I think about this, the less I like it. Could we append a newline and block_indent instead of using a block comment in the expression case? It feels a bit irregular atm.

Comment (marcusklaas) [inline src/lists.rs]: I agree it's not pretty, but I am afraid it will be a bit irregular in any case. If I understand correctly, we have to decide between 
```rust
(item1, // Comment 1
 item2, // Comment 2
 item3, /* Comment 3 */)
```
and
```rust
(item1, // Comment 1
 item2, // Comment 2
 item3, // Comment 3
)
```
Personally, I don't believe the latter is always better than the first, but I can get behind it if that's what is decided on.

Comment (nrc) [inline src/lists.rs]: I don't want to worry too much about this - it should be a total edge case exactly because there is no way to format it nicely. Another alternative might be to change all post-comments to /* */. Or to only do that if there is a final comment? Any way we do this is kind of inconsistent. I guess I prefer a little inconsistency in the punctuation rather than in the comment style (and the code should be simpler), but I don't think it is too important.

File: src/lists.rs (hunk 11 above)
Comment (marcusklaas) [inline src/lists.rs]: This was actually a bug! :fearful:

File: src/lists.rs (hunk 10 above)
Comment (marcusklaas) [inline src/lists.rs]: Alright. I've left it as it is for now. We can revisit this later, if you do not mind.

**Comment by marcusklaas** [comment]
Nice! :confetti_ball:

Thank you for your guidance and patience in this, @nrc! I learned a lot doing it!

**Comment by nrc** [comment]
A pleasure; thanks for the PR!
